core/thread/auth.go

```diff
@@ -0,0 +1,155 @@
+package thread
+
+import (
+	"context"
+	"encoding"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/gogo/status"
+	"github.com/grpc-ecosystem/go-grpc-middleware/util/metautils"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"google.golang.org/grpc/codes"
+)
+
+type Identity interface {
+	Sign([]byte) ([]byte, error)
+	GetPublic() crypto.PubKey
+}
+
+type Auth struct {
+	pk  crypto.PubKey
+	sig []byte
+}
+
+type Method byte
+
+const (
+	CreateThread  Method = 0x01
+	AddThread            = 0x02
+	GetThread            = 0x03
+	PullThread           = 0x04
+	DeleteThread         = 0x05
+	AddReplicator        = 0x06
+	CreateRecord         = 0x07
+	AddRecord            = 0x08
+	GetRecord            = 0x09
+	Subscribe            = 0x0A
+)
+
+func NewAuth(id Identity, method Method, args ...encoding.BinaryMarshaler) (auth Auth, err error) {
+	if id == nil {
+		return
+	}
+	msg, err := marshalBinary(method, args)
+	if err != nil {
+		return
+	}
+	sig, err := id.Sign(msg)
+	if err != nil {
+		return
+	}
+	return Auth{
+		pk:  id.GetPublic(),
+		sig: sig,
+	}, nil
+}
+
+func NewAuthFromMD(ctx context.Context) (auth Auth, err error) {
+	val := metautils.ExtractIncoming(ctx).Get("authorization")
+	if val == "" {
+		return
+	}
+	val = strings.TrimPrefix(val, "Signature ")
+	parts := strings.SplitN(val, ",", 2)
+	if len(parts) < 2 {
+		return auth, status.Error(codes.Unauthenticated, "Bad authorization string")
+	}
+	pkstr := strings.TrimPrefix(parts[0], "keyId=")
+	pkb, err := base64.StdEncoding.DecodeString(pkstr)
+	if err != nil {
+		return auth, status.Error(codes.Unauthenticated, "Bad authorization string")
+	}
+	pk, err := crypto.UnmarshalPublicKey(pkb)
+	if err != nil {
+		return auth, status.Error(codes.Unauthenticated, "Bad public key")
+	}
+	sigstr := strings.TrimPrefix(parts[1], "signature=")
+	sig, err := base64.StdEncoding.DecodeString(sigstr)
+	if err != nil {
+		return auth, status.Error(codes.Unauthenticated, "Bad authorization string")
+	}
+	return Auth{
+		pk:  pk,
+		sig: sig,
+	}, nil
+}
+
+func (a Auth) String() string {
+	if a.pk == nil {
+		return ""
+	}
+	pkb, err := crypto.MarshalPublicKey(a.pk)
+	if err != nil {
+		panic(err)
+	}
+	keystr := base64.StdEncoding.EncodeToString(pkb)
+	sigstr := base64.StdEncoding.EncodeToString(a.sig)
+	return fmt.Sprintf("Signature keyId=%s,signature=%s", keystr, sigstr)
+}
+
+func (a Auth) Verify(method Method, args ...encoding.BinaryMarshaler) (crypto.PubKey, error) {
+	if a.pk == nil {
+		return nil, nil
+	}
+	msg, err := marshalBinary(method, args)
```

Comment:
On the verification end, we can create the exact same message _from any of the APIs_... not just the gRPC ones since the message does not involve protos.